### PR TITLE
Create a separate utility for reporting.

### PR DESCRIPTION
**Pull request by adelez**

### CODE REVIEW

File: tools/run_tests/jobset.py [hunk 1]
```diff
@@ -229,15 +226,12 @@ def state(self, update_cache):
       self._tempfile.seek(0)
       stdout = self._tempfile.read()
       filtered_stdout = _filter_stdout(stdout)
```
Comment (jtattermusch) [inline tools/run_tests/jobset.py]: I think the logic for escaping (_filter_stdout) and truncating the stdout  should be part of the report generation. Also the same logic 
```
filtered_stdout = _filter_stdout(stdout)
filtered_stdout = filtered_stdout[-128:]
```

is used multiple times in run_tests.py and run_interop_tests.py, and this PR sound like a good opportunity to do some cleanup.

Also, the the TODO is out of date and we can delete it. And we should probably return last 8KB of stdout, 128 bytes is not enough to display any meaningful troubleshooting info (like stack trace).

File: tools/run_tests/run_interop_tests.py [hunk 2]
```diff
@@ -740,22 +620,20 @@ def render_html_report(client_langs, server_langs, resultset,
...
-                                       maxjobs=args.jobs, xml_report=testsuite)
+                                       maxjobs=args.jobs)
   if num_failures:
     jobset.message('FAILED', 'Some tests failed', do_newline=True)
   else:
     jobset.message('SUCCESS', 'All tests passed', do_newline=True)
 
-  tree = ET.ElementTree(root)
-  tree.write('report.xml', encoding='UTF-8')
+  # Generate XML report.
```
Comment (jtattermusch) [inline tools/run_tests/run_interop_tests.py]: you can remove the comment, I think the method name speaks for itself.

File: tools/run_tests/generate_reports.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
+# Copyright 2015, Google Inc.
```
Comment (jtattermusch) [inline tools/run_tests/generate_reports.py]: Nit: I'd be leaning towards some other module name than "generate_reports" which looks more like a function name. 
Just "report" sounds good (if it doesn't collide with some standard python module name). "report_generator" or "report_utils" also sound good.

File: tools/run_tests/jobset.py (hunk 1 above)
Comment (adelez) [inline tools/run_tests/jobset.py]: I'm afraid without truncating, the resultset object may be unnecessarily too big to hold in memory if you run all the tests many times.

File: tools/run_tests/run_interop_tests.py (hunk 2 above)
Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: Done

File: tools/run_tests/generate_reports.py (hunk 3 above)
Comment (adelez) [inline tools/run_tests/generate_reports.py]: Done.

File: tools/run_tests/jobset.py (hunk 1 above)
Comment (jtattermusch) [inline tools/run_tests/jobset.py]: Good point. Can we at least do some cleanup and unify the truncating logic and perhaps add a todo?
One solution that sounds reasonable to me:
1.  add the max_result_size in the job/jobspec/jobset/.. configuration, truncation should be done upon job result creation.
2. the exact way of filtering output depends on the report output format (xml, html, ...) so filtering should be done in the report module, not here.

File: tools/run_tests/run_interop_tests.py (hunk 2 above)
Comment (jtattermusch) [inline tools/run_tests/run_interop_tests.py]: FYI, changes are not visible here yet.

Comment (adelez) [inline tools/run_tests/run_interop_tests.py]: Sorry forgot to push. Done now.

File: tools/run_tests/jobset.py (hunk 1 above)
Comment (adelez) [inline tools/run_tests/jobset.py]: Done.

File: tools/run_tests/jobset.py [hunk 4]
```diff
@@ -34,15 +34,14 @@
...
-import string
 import subprocess
 import sys
 import tempfile
 import time
-import xml.etree.cElementTree as ET
 
 
 _DEFAULT_MAX_JOBS = 16 * multiprocessing.cpu_count()
+_MAX_RESULT_SIZE = 128
```
Comment (jtattermusch) [inline tools/run_tests/jobset.py]: please lift to at least 8192

File: tools/run_tests/jobset.py [hunk 5]
```diff
@@ -224,20 +213,19 @@ def start(self):
 
   def state(self, update_cache):
     """Poll current state of the job. Prints messages at completion."""
+    self._tempfile.seek(0)
+    stdout = self._tempfile.read()
+    # TODO: looks like jenkins master is slow because parsing the junit
```
Comment (jtattermusch) [inline tools/run_tests/jobset.py]: pls remove this TODO, it is not up to date.

Comment (jtattermusch) [inline tools/run_tests/jobset.py]: we already know that the reason for jenkins being slow was different.

File: tools/run_tests/jobset.py [hunk 6]
```diff
@@ -224,20 +213,19 @@ def start(self):
 
   def state(self, update_cache):
     """Poll current state of the job. Prints messages at completion."""
+    self._tempfile.seek(0)
+    stdout = self._tempfile.read()
+    # TODO: looks like jenkins master is slow because parsing the junit 
+    # results XMLs is not implemented efficiently. This is an experiment to 
+    # workaround the issue by making sure results.xml file is small enough.
+    if len(stdout) > _MAX_RESULT_SIZE:
```
Comment (jtattermusch) [inline tools/run_tests/jobset.py]: wouldn't just self.result.message = stdout[-_MAX_RESULT_SIZE:] without  the if - else suffice?

**Comment by jtattermusch** [comment]
LGTM after addressing the three nits.

File: tools/run_tests/jobset.py (hunk 4 above)
Comment (adelez) [inline tools/run_tests/jobset.py]: Done

File: tools/run_tests/jobset.py (hunk 5 above)
Comment (adelez) [inline tools/run_tests/jobset.py]: Done

File: tools/run_tests/jobset.py (hunk 6 above)
Comment (adelez) [inline tools/run_tests/jobset.py]: Done

**Pull request merged**